src/Simulation/Native/src/simulator/simulator.hpp

```diff
@@ -71,6 +71,12 @@ class Simulator : public Microsoft::Quantum::Simulator::SimulatorInterface
         return psi.isclassical(q);
     }
 
+    std::pair<bool, bool> is_classical_or_entangled(logical_qubit_id q)
+    {
+        recursive_lock_type l(mutex());
+        return psi.is_classical_or_entangled(q);
+    }
```

Comment:
FYI:
> recursive_lock_type l(mutex());

I saw this fragment a number of times in the existing code and I don't understand what this fragment does. It locks not the _existing mutex_, but a mutex created on the fly (new mutex created every time the function is entered). That is why the fragment does not protect from re-entrancy and does not synchronize the access during concurrency.

Can you please explain to me what this fragment does?

Reply:
Interesting... I hadn't noticed this until you pointed it out, so I'm glad you highlighted it! To call the constructor for a mutex this would actually have to be `std::mutex()` where in this case it is instead intended to invoke the (rather poorly named) member function `mutex()` from simulationinterface.hpp: https://github.com/microsoft/qsharp-runtime/blob/b908e8d62b2c67df6999546dd45425f6e6a45db1/src/Simulation/Native/src/simulator/simulatorinterface.hpp#L127-L130
That will get the pointer to the member mutex for this class. That said, its possible that depending on how this is compiled and if any header files have a `using namespace std;` in them, it could change how this call works and cause it to behave exactly as you describe. That could even be the result of the intermitten failures that we've seen in functions in this class when running multithreaded tests. To make the use clearer and explicitly avoid this possible source of error, I'll update all these call points to use `this->mutex()` instead. Maybe that will even fix our intermittent issue!

Reply:
:-DDDDDDDDDDDDDDDDDDDDDDD
Incredible!
Thank you very much!

Reply:
Can we rename this function so that there's no confusion now AND in future code?

Reply:
Might as well... the naming conventions in this simulator code are a bit odd and seem to prioritize terseness over readability or maintainability, so I have no problem making that change. In that case, I should be able to remove the explicit use of the `this` pointer.